enhancements/agent-installer/agent-based-installer.md

```diff
@@ -0,0 +1,279 @@
+---
+title: agent-based-installer
+authors:
+  - "@zaneb"
+reviewers:
+  - "@avishayt"
+  - "@celebdor"
+  - "@hardys"
+  - "@lranjbar"
+  - "@mhrivnak"
+  - "@patrickdillon"
+  - "@pawanpinjarkar"
+  - "@romfreiman"
+  - "@rwsu"
+  - "@sdodson"
+approvers:
+  - "@dhellmann"
+creation-date: 2022-03-03
+last-updated: 2022-03-03
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - https://issues.redhat.com/browse/AGENT-9
+---
+
+# Agent-based Installer
+
+## Summary
+
+The agent-based installer is an installation method that will use a bootable,
+ephemeral installer image running on the hosts that are to become part of the
+cluster. The user will generate the image using a command-line tool. The image
+itself will contain components (such as
+[assisted-service](https://github.com/openshift/assisted-service/#readme) and
+[assisted-agent](https://github.com/openshift/assisted-installer-agent#readme))
+of the Assisted and Multi-Cluster Engine (MCE) installation methods. The
+installer will be usable in both a fully-automated workflow, where
+configuration is provided upfront, and an interactive workflow where the user
+can interact with a GUI similar to the hosted Assisted installation service.
+
+## Motivation
+
+Every on-premises user has a first cluster ("Cluster 0"), and many of them will
+be installed in a disconnected environment. While powerful tools exist for
+creating multiple on-premises clusters (Hive, HyperShift, MCE), they all run in
+a cluster themselves, which must somehow be provisioned without these tools.
+
+Each of the current installation options has limitations that make it
+unpalatable to a subset of users. IPI is highly opinionated, and suffers from
+providing limited visibility of configuration problems in on-premises
+environments. UPI exposes users to the full complexity of OpenShift to the user
+with little guidance. The (hosted) Assisted service is not available for
+IPv6-only or disconnected environments. For some users, none of these options
+is optimal.
+
+On-premises users need a simple way to deploy their first cluster, even in a
+disconnected environment, using their own automation to boot hosts.
+
+### Goals
+
+- Install clusters in fully-disconnected environments.
+- Make use of assisted installation components to pre-validate the installation
+  environment.
+- Require no machines in the cluster environment other than those that are to
+  become part of the cluster.
+- Install clusters with single-node, compact, and high-availability topologies.
+- Be agnostic to the tools used to provision (virtual) machines.
+- Support install configurations with platform `none`, `baremetal`, or
+  `vsphere`.
+
+### Non-Goals
+
+- Replace the IPI installer in any capacity.
+- Automate booting of the image on (virtual) machines.
+- Support install configurations for cloud-based platforms.
+- Provide an interactive API to drive automated deployments.
```

Comment:
This document does not describe how we would run an API service that an orchestration tool would use to trigger provisioning because we have requirements that have led us to conclude that the tool can't work that way. Some of the details I've suggested adding in my comment below (https://github.com/openshift/enhancements/pull/1053/files#r821861322) will make that clearer.

It's possible this bit of information would fit better in the proposal or details section (maybe in the rough overview of the workflow that @julienlim asked about), since it's a conclusion rather than a starting condition for the design.

Reply:
I included it in the non-goals because it isn't a goal: no user has requested it and nobody afaik has articulated a need for it. If it _were_ a goal then we would need to adjust the design and/or compromise on other goals. Nevertheless, it comes up a lot so it is worth being explicit about this as a starting condition.

This doesn't necessarily mean that it is an anti-goal, but the _conclusion_ is that due to the other constraints we have we're not going to have one.